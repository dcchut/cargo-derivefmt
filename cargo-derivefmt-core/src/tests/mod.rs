macro_rules! test {
    ($contents: expr) => {
        let mut source = textwrap::dedent($contents);
        crate::modify_source(&mut source);
        insta::assert_snapshot!(source);
    };
}

#[test]
fn test_derive_missing_comma() {
    test!(
        r#"
        #[derive(Debug, Clone, Default Hash)]
        struct S;
    "#
    );
}

#[test]
fn test_derive_double_comma() {
    test!(
        r#"
        #[derive(Debug, Clone, Default,, Apples, Hash)]
        struct T;
    "#
    );
}

#[test]
fn test_derive_ordering() {
    test!(
        r#"
        #[derive(Eq, Ord, PartialOrd, Copy, Clone, Debug, PartialEq, Hash)]
        struct Wrapped<T>(T);
    "#
    );
}

#[test]
fn test_derive_ordering_qualified() {
    test!(
        r#"
        #[derive ( PartialEq, Copy, PartialOrd, std :: fmt :: Debug, Hash, Eq, Ord )]
        struct SillyBugger;
    "#
    );
}

#[test]
fn test_multiline_derives() {
    test!(
        r#"
        #[derive(
            core::cmp::PartialEq,
            core::clone::Clone,
            core::marker::Copy,
            core::fmt::Debug,
            core::hash::Hash,
            core::default::Default,
            core::cmp::Eq,
            core::cmp::Ord,
            core::cmp::PartialOrd,
        )]
        struct Core;

        #[derive(
            std::fmt::Debug,
            std::clone::Clone,
            std::marker::Copy,
            std::cmp::PartialEq,
            std::default::Default,
            std::cmp::Eq,
            std::hash::Hash,
            std::cmp::Ord,
            std::cmp::PartialOrd,
        )]
        struct Std;

        "#
    );
}

#[test]
fn test_derive_comments() {
    test!(
        r#"
        #[derive(
        /* ---------- Some really important comment that just had to go inside the derive --------- */
        Debug,
        Clone, // And what about this?
        Eq, PartialEq,
        )]
        struct Foo {
            a: i32,
            b: T,
        }
        "#
    );
}
