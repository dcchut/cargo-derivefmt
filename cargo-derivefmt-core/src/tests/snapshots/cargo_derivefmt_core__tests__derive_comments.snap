---
source: cargo-derivefmt-core/src/tests/mod.rs
expression: source
---

#[derive(
Clone,
/* ---------- Some really important comment that just had to go inside the derive --------- */
Debug, // And what about this?
Eq, PartialEq,
)]
struct Foo {
    a: i32,
    b: T,
}

